Sprint2

seb4ever opened the pull request:
We have also made progress in the branch here: https://github.com/paulaSiemens/newsRoom/tree/header-navbar-skeleton, which needs to be merged with sprint2.

Commit:
initial commit, boiler plate react

Commit:
added yearn install to readme file

Commit:
App scaleton

Commit:
adjusted root folder according to Bjørn's recommendations

Commit:
Update README.md

Commit:
Installed parse

Commit:
Added routing and navbar with basic login and signup forms according to Dansk in town video by Mircea

Commit:
Added connection with database and additional functionality

Commit:
add handler for login

Evaise20 committed:
markup added

Commit:
fix selection screen to fetch all ideas from database

Commit:
add upload image file to SubmitIdea, add display idea image in Selection

Commit:
fix update Header when logging in, add logout functionality, add active tab highlight

Commit:
added header styling and navbarLeft component

bootstrap styling overrides in Header.css and NavbarLeft.css

Commit:
added maincontainer

maincontainer with flex display
container: navbarLeft and MainContent

Commit:
add signup functionality

Commit:
added icon imgs

Commit:
sidenav details

sidenav styling, router/links and new component files

Commit:
fix restricted access routes for logged in users (default: '/') vs non-logged in users (dfault: '/login'

bwesth commented on README.md:
```diff
@@ -1,2 +1,78 @@
 # newsRoom
 TID Group 13
+
+# Getting Started with Create React App
+
+This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).
```
If you have 5 minutes it would be nice with your own readme file :)

bwesth commented on src/App.js:
```diff
@@ -0,0 +1,38 @@
+      <Header />
+      <BrowserRouter>
+        <Routes>
+          <Route path="/signup" element={!Parse.User.current( ) ? <Signup /> : <Navigate to='/'/>} />
+          <Route path="/login" element={!Parse.User.current( ) ? <Login /> : <Navigate to='/'/>} />
+          <Route path="/submitIdea" element={Parse.User.current( ) ? <SubmitIdea /> : <Navigate to='/login'/>} />
+          <Route path="/selection" element={Parse.User.current( ) ? <Selection /> : <Navigate to='/login'/>} />
```
That's one way to do it.
Try to see if you can limit your identical occurrences of conditional logic.
If you have the same conditions, can you somehow group them?

bwesth commented on src/App.js:
```diff
@@ -0,0 +1,38 @@
+          <Route path="/selection" element={Parse.User.current( ) ? <Selection /> : <Navigate to='/login'/>} />
+        </Routes>
+      </BrowserRouter>
+      <br />
+      <br />
+      <br />
+      <br />
+      <br />
+      <br />
+      <br />
+      <br />
```
https://preview.redd.it/o5glo0ipdn951.jpg?auto=webp&s=6706938b2424c77edb9c54ec792c96006d9a4a37
:)

Maybe some margin or padding instead?

bwesth commented on src/components/Footer.js:
```diff
@@ -0,0 +1,9 @@
+
+function Footer() {
+    return (
+        <footer>This is the footer</footer>
```
Nice! :D

bwesth commented on src/components/Header.js:
```diff
@@ -0,0 +1,55 @@
+}, []);
+
+  function handleLogoutAttempt(e) {
+      alert("You have logged out");
+      setLoggedIn(false);
+      Parse.User.logOut();
+      window.location.href="/";
```
Sure you want to setLoggedIn before Parse.user.logOut?
Maybe the parse logout returns an unforeseen error, and since it is asynchronous, your system will be out of sync.
This could be a non-issue in this case, but it is something you should consider always when using APIs.
Good rule is to await the API call, and then do your own thing. Catch in case of error and make sure to separate "on success" logic from "on failure" logic.

bwesth commented on src/components/Selection.js:
```diff
@@ -0,0 +1,35 @@
+import { useEffect, useState } from "react";
+import Parse from "parse";
```
Here and other places where you import Parse:
Consider making a separate .js file with all your parse API functionality together.
could call it api.js :D

bwesth commented on src/components/Selection.js:
```diff
@@ -0,0 +1,35 @@
+    });
+  }, []);
+
+  if (!selection) {
+    return <p>Loading...</p>;
+  }
+
+  return (
+    <>
+       <br />
+      <h1>Selection</h1>
+      <br />
+      <br />
+      <ul>
+      {selection.map((idea, i) => (
+       <li key={i}><img src={ idea.get("image").url()} alt="illustration expressing the idea"  /><br/><b>{idea.get("title")}</b><br/>{idea.get("description")}<hr/></li>
+      ))}
+      </ul>
+    </>
+  );
+}
```
There are differing opinions on having 2 return statements.
I know that Sara dislikes them, however. I can read this just fine, but there is an argument for readability supporting 1 return statement.

You can move the if down into the main return block with a ternary statement (a sexy version of "if-else").

bwesth replied:
I realize how this reads, so just be absolutely clear:
I am **not** saying that I can read this fine, but Sara cannot.
I mention Sara's opinion because I respect her, and would like to defer to her on this matter because I think she has **proper reasons** for disliking 2 return statements, whereas I base my own _lack-of-opinion-tilting-to-the-side-of-dislike_ merely on a feeling.

In either case, it's not as much an error on your part, as something I want you to be aware of.

mircealungu replied:
Just to add to the discussion: I often use early returns too because personally I find the code easier to parse visually that way.

bwesth commented on src/components/Signup.js:
```diff
@@ -0,0 +1,62 @@
+  const navigate = useNavigate();
+
+   async function handleSignupAttempt(e) {
+    e.preventDefault();
+
```
Consider why you need to preventDefault. What is "default" in this case?

bwesth commented on src/components/Login.js:
```diff
@@ -0,0 +1,52 @@
+    const user = new Parse.User();
+    user.setPassword(password);
+    user.setUsername(email);
+    user.logIn().then((loggedInUser) => {
+      navigate("/")
+      window.location.reload(false);
+    });
```
What happens in the case of an exception? Anyone catching that?

bwesth reviewed (approved):
Generally, you know what you're doing.
Good job! :) 
I suggest, before your ITU days are over, that you practice making your own components.
I realize time is limited right now, and Bootstrap is a great framework, but making components is like practising your s's and o's in when you were in primary school. If you never make them yourselves you'll always only be able to use someone else's framework. For now, no problemo.

Another thing: I think you may have style.css on gitignore? I can see it linked in public/index.html, but I don't see it anywhere. Obviously, I can't see all your cool styling as a consequence.

Commit:
initial commit

Commit:
Merge branch 'sprint2-mergeCopy' into header-navbar-skeleton

Commit:
removed <br />'s

Commit:
fix handleLogoutAttempt

Commit:
header styling fix

Commit:
moved logout to icon

Commit:
moved navigation to sidebar

Commit:
moved loggedin browserRouter to MainContent

Commit:
moved routing back to app

Commit:
homepage set to assigned

changed navigation after login to new assigned window,
changed navigation on brand click to assigned window (when logged in)

Commit:
added empty rchived and submitted windows

Commit:
Merge branch 'sprint2' into header-navbar-skeleton

Commit:
Merge pull request #3 from paulaSiemens/header-navbar-skeleton

Header navbar skeleton
